netket/abstract_driver.py

```diff
@@ -0,0 +1,281 @@
+            for i in range(0, step):
+                dp = self._forward_and_backward()
+                if i is 0:
+                    yield self.step_count
+
+                self.update_parameters(dp)
+
+    def advance(self, steps=1):
+        """
+        Performs `steps` optimization steps.
+
+        :param steps: (Default=1) number of steps
+        """
+        for _ in self.iter(steps):
+            pass
+
+    def run(
+        self,
+        output_prefix,
+        n_iter,
+        obs=None,
+        save_params_every=50,
+        write_every=50,
+        step_size=1,
+        show_progress=True,
+    ):
+        """
+        Executes the Monte Carlo Variational optimization, updating the weights of the network
+        stored in this driver for `n_iter` steps and dumping values of the observables `obs`
+        in the output. The output is a json file at `output_prefix`, overwriting files with
+        the same prefix.
+
+        Args:
+            :output_prefix: The prefix at which json output should be stored (ignored if logger
+              is provided).
+            :n_iter: the total number of iterations
+            :obs: An iterable containing all observables that should be computed
+            :save_params_every: Every how many steps the parameters of the network should be
+            serialized to disk (ignored if logger is provided)
+            :write_every: Every how many steps the json data should be flushed to disk (ignored if
+            logger is provided)
+            :step_size: Every how many steps should observables be logged to disk (default=1)
+            :show_progress: If true displays a progress bar (default=True)
+        """
+        if obs is None:
+            # TODO remove the first case after deprecation of self._obs in 3.0
+            if len(self._obs) is not 0:
+                obs = self._obs
+            else:
+                obs = {}
+
+        logger = _JsonLog(output_prefix, save_params_every, write_every)
+
+        # Don't log on non-root nodes
+        if self._mynode is not 0:
+            logger = None
+
+        with tqdm(
+            self.iter(n_iter, step_size), total=n_iter, disable=not show_progress
+        ) as itr:
+            for step in itr:
+                # if the cost-function is defined then report it in the progress bar
+                if self._stats is not None:
+                    itr.set_postfix_str(self._stats_name + "=" + str(self._stats))
+
+                obs_data = self.estimate(obs)
+
+                if self._stats is not None:
+                    obs_data[self._stats_name] = self._stats
+
+                log_data = tree_map(obs_stat_to_dict, obs_data)
+
+                if logger is not None:
+                    logger(step, log_data, self.machine)
+
+        # flush at the end of the evolution so that final values are saved to
+        # file
+        if logger is not None:
+            logger.flush(self.machine)
```

Comment:
TLDR: I sort of agree with you.
--
My intent for this PR was to do a PR with no breaking changes.
I have another PR sort-of ready with breaking changes (the argument order of `vmc.run`, for example). We can do this in the future PR.

The reason is that... JSON (in my opinion) is a shitty format. You cannot append to it. So at every step you are rewriting the full file, which eventually becomes very costly.  
Rather than flushing to file at every iteration (which you can do, just by passing the correct `save_every` flag), we could add some logic to JsonLogger to keep flushing as long as writing to file does not take more than XX% of total computational time.

Reply:
Ok, then let's leave it in the current state for this PR.

I agree with JSON being sub-optimal for scientific data. Still, as an easy to read fallback format I think it beats ad-hoc CSV files or similar text output as it is at least structured.

For my own code I currently use a format where every line of the logfile is a JSON object (solves the appending problem and is still easy to process) for the high-level log info and [zarr](https://zarr.readthedocs.io/en/stable/) for larger array data (e.g., the machine parameters, correlators between all pairs of sites). That works quite well with `vmc.iter`.

Reply:
I was not aware of Zarr.
This might be nice.